src/databricks/labs/ucx/assessment/azure.py

```diff
@@ -188,45 +137,77 @@ def _safe_get_cluster_policy(self, policy_id: str) -> Policy | None:
+        For each application id identified, retrieve the relevant storage account, and corresponding tenant id
+        """
+        spn_list = []
+        spn_application_id, secret_scope, secret_key, tenant_id, storage_account = None, "", "", "", ""
+        matching_spn_app_id_keys = [key for key in config.keys() if "fs.azure.account.oauth2.client.id" in key]
+        if len(matching_spn_app_id_keys) > 0:
+            for spn_app_id_key in matching_spn_app_id_keys:
+                # retrieve application id of spn
+                spn_application_id, secret_scope, secret_key = self._get_key_from_config(spn_app_id_key, config)
+                if spn_application_id is None:
                     continue
-                temp_list = self._get_azure_spn_list(pipeline_config)
-                if temp_list:
-                    azure_spn_list_with_data_access_from_pipeline += temp_list
-        return azure_spn_list_with_data_access_from_pipeline
-
-    def _assess_service_principals(self, relevant_service_principals: list):
-        for spn in relevant_service_principals:
-            spn_info = AzureServicePrincipalInfo(
-                application_id=spn.get("application_id"),
-                secret_scope=spn.get("secret_scope"),
-                secret_key=spn.get("secret_key"),
-                tenant_id=spn.get("tenant_id"),
-                storage_account=spn.get("storage_account"),
-            )
-            yield spn_info
 
-    def snapshot(self) -> Iterable[AzureServicePrincipalInfo]:
-        return self._snapshot(self._try_fetch, self._crawl)
+                # retrieve storage account configured with this spn
+                storage_account_matched = re.search(STORAGE_ACCOUNT_EXTRACT_PATTERN, spn_app_id_key)
+                if storage_account_matched:
+                    storage_account = storage_account_matched.group(1).strip(".")
+                    tenant_key = "fs.azure.account.oauth2.client.endpoint." + storage_account
+                else:
+                    tenant_key = "fs.azure.account.oauth2.client.endpoint"
 
-    def _try_fetch(self) -> Iterable[AzureServicePrincipalInfo]:
-        for row in self._fetch(f"SELECT * FROM {self._schema}.{self._table}"):
-            yield AzureServicePrincipalInfo(*row)
+                # retrieve tenant id of spn
+                matching_tenant_keys = [key for key in config.keys() if re.search(tenant_key, key)]
+                if len(matching_tenant_keys) == 0 or matching_tenant_keys[0] is None:
+                    tenant_id = ""
+                else:
+                    client_endpoint_list = self._get_key_from_config(matching_tenant_keys[0], config)[0].split("/")
+                    if len(client_endpoint_list) == CLIENT_ENDPOINT_LENGTH:
+                        tenant_id = client_endpoint_list[3]
+
+                # add output to spn list
+                spn_list.append(
```

Comment:
i meant creating the `AzureServicePrincipalInfo` from here directly, without using intermediate dictionary structure. can you change this place and the other couple of places where we construct this dict?